rfd/0009-viewer-app.md

```diff
@@ -0,0 +1,76 @@
+---
+authors: Mathew (wrm932@gmail.com)
+state: draft
+---
+
+# Required Approvers
+* Engineering @pierrebeaucamp && @jimbishopp && @michellescripts && @mcbattirola
+
+# RFD 0009 - RFDs
+
+## What
+
+Create a directory viewer similar to Github and Google drive. 
+
+
+## Why
+
+A pratical exam that tests the potential applicant on their fullstack software knowledge. 
+
+## Details
+
+Each RFD is stored in a markdown file under
+https://github.com/gravitational/teleport/tree/master/rfd and has a unique
+number.
+
+The directory viewer should be allow the user to navigate a directory structure by clicking on files or folders. A bread crumb navigation should inform them of where they currently are and allow them to navigate back to their previously clicked directory location. The user should also be able to filter out a list of files using a search box. Search/filtering will be done client side. 
+
+ In addition, user authentication should be implemented so only logged in users will be able to access the viewer. If a user is unauthenticated they should be redirected back to the login form. However, if given a direct url that points to a specific directory. The application should be able to remember it and redirect the unauthenticated user to the proper url after successful authentication. 
+
+## API 
+
+API should return a directory structure json response 1-level deep at a time. Subsequent calls to this api with added directory parameters should return the appropriate json response. 
+
+```json
+GET /directory/[...slug]
+{
+  name: "teleport",
+  size: 0, // bytes
+  type: "dir",
+  items: [
+    {
+      name: "lib",
+      size: 0, // bytes
+      type: "dir",
+      items: [
+        {
+          name: "teleport.go",
+          size: 320000, // bytes
+          type: "file",
+        },
+        {
+          name: "test.go",
+          size: 3320000, // bytes
+          type: "file",
+        },
+      ],
+    },
+    {
+      name: "README.md",
+      size: 4340000, // bytes
+      type: "file",
+    },
+  ],
+}
+
+```
+
+### Security
```

Comment:
Do you think CSRF is a concern in this app?

Reply:
Yes I think so! And maybe I should be have been more aware that a possible attack vector would be a `POST` request to a login endpoint.